Implement exponentially scaled slider widget

netnazgul opened the pull request:
Closes #14696.

[rationale](https://www.dr-lex.be/info-stuff/volumecontrols.html)

Constant values a = 0.001 and b = 6.908 for f(x) = a*e^(b*x) rescale the curve to ~60dB span.

![image](https://user-images.githubusercontent.com/1614714/34836746-ece8f18a-f709-11e7-8136-745709343a81.png)

Custom mods that use different settings implementation will need to be updated with ExpSliderWidget for appropriate sliders.

netnazgul committed:
Implement a slider widget for volume control that returns an exponentially scaled value

Unit158 commented on OpenRA.Mods.Common/Widgets/ExpSliderWidget.cs:
```diff
@@ -0,0 +1,76 @@
+
+		float ExpFromLinear(float x)
+		{
+			return (float)(x == MinimumValue ? MinimumValue : 1.0e-3 * Math.Exp(6.908 * x)).Clamp(MinimumValue, MaximumValue);
```
Why not just drop the ternary operator if you're going to clamp it anyway?

netnazgul replied:
because e^x does not produce 0 at x=0 (the result of above function is 0.001), meaning you are not able to mute sound. 
Clamp was actually meant for approximation before, because it turned out slider value is derived from mouse position relative to slider widget BEFORE clamping, meaning it produced negative values. But still that same e^x produces values slightly higher than 1.0 on x=1.0 (1.0024 for the above function)

Unit158 replied:
My point stays, the ternary should go. If you want it to round to zero at the bottom, then make it so.

netnazgul replied:
So how do I round it to minimum at the bottom without ternary?

Unit158 replied:
youre completely right and im an idiot, again. Just change the equals to a less than or equals.

netnazgul replied:
Hmm... Yeah, I was assuming the span of 0..1 for Min/Max. Will think of something about this

pchote replied:
This line is a bit of a monster.  Can you please reformat it to:
```csharp
if (x == MinimumValue)
    return MinimumValue;

return 1.0e-3 * Math.Exp(6.908 * x)).Clamp(MinimumValue, MaximumValue);
```

It might also be a good idea to pull those constants out to fields so that they can be overridden from yaml.

Unit158 commented on OpenRA.Mods.Common/Widgets/ExpSliderWidget.cs:
```diff
@@ -0,0 +1,76 @@
+
+		float LinearFromExp(float x)
+		{
+			return (float)(x == MinimumValue ? MinimumValue : Math.Log(x / 1.0e-3) / 6.908).Clamp(MinimumValue, MaximumValue);
```
same here

netnazgul replied:
here the ternary can be dropped, yes. x=0.0 will be clamped (function produces -INF).

fruestueck commented:
I've tried it and it sounds good : )
Only the lower part (first quarter of the slider) felt like there was almost no change. Could be me too.

netnazgul commented:
@fruestueck -60dB is pushing it on the dynamic range of some mainstream sound devices, especially if you do not wind them up at 100% output at the time. I didn't hear anything in my headphones as well on the lower settings. As we can't predict the system user uses in his sound path, we'd better have a bit more of excess control for that. But with this implemented there is at least greater control over more quiet parts of configuration, keeping close to the same scale for general use (which for me is around -20dB..-8dB)

Unit158 commented on OpenRA.Mods.Common/Widgets/ExpSliderWidget.cs:
```diff
@@ -0,0 +1,76 @@
+
+		protected override int PxFromValue(float x)
+		{
+			return (int)(0.5f * RenderBounds.Height + (RenderBounds.Width - RenderBounds.Height) * (LinearFromExp(x) - MinimumValue) / (MaximumValue - MinimumValue));
```
Why multiply by 0.5 when you divide by 2 in the Draw function? Not a deal breaker, but I'd personally prefer division by 2.

netnazgul replied:
This is the exact copy of the same methods as in SliderWidget:
https://github.com/OpenRA/OpenRA/blob/fadcfa0828bee9d00fcfdaccca92970be70a4926/OpenRA.Mods.Common/Widgets/SliderWidget.cs#L87-95

pchote commented on OpenRA.Mods.Common/Widgets/ExpSliderWidget.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	public class ExpSliderWidget : SliderWidget
```
Style nit: Can we please call this `ExponentialSliderWidget`?  Exp is ambiguous with "Experience".

pchote commented:
@netnazgul do you plan on getting back to this?

netnazgul commented:
Yes, sorry, couldn't find enough time to fix this. Will look through the next week.
One issue raised through the discussion above - I think it will be hard to apply the proper rescaling if MinValue/MaxValue are not set to 0/1.

netnazgul commented:
- Renamed ExpSliderWidget into ExponentialSliderWidget
- Moved exponential coefficients to explicit variables
- Removed ternary ops; simple (x <= 0) check is left to have f(0) = 0
- Modified rescaling applications so that MinimumValue and MaximumValue are correctly applied

netnazgul commented:
oh wow, tried to fix the AUTHORS tab instead of spaces using built-in github editor and it failed, will fix and squash it later today

netnazgul commented:
Done, forgot about this...

Unit158 commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -124,6 +124,17 @@ static void BindSliderPref(Widget parent, string id, object group, string pref)
 			ss.OnChange += x => field.SetValue(group, x);
 		}
 
+		static void BindExpSliderPref(Widget parent, string id, object group, string pref)
```
Not in scope for this PR, but these binding functions should be factored.

abcdefg30 reviewed (dismissed):
This crashes when opening the music player tab saying the game can't convert (cast) `SliderWidget` to `ExponentialSliderWidget`.

pchote commented:
Does someone want to take this over, or shall we close it?

netnazgul commented:
I'll update it on friday/saturday if you still want to push it.

In fact I've come to conclusion, that implementing this with the exponential formula as is doesn't provide much benefit to what I've initially wanted here - expanding the control area around where the most volume tweaks are done. Basically, what it does now is moving the mentioned area from the left part of the slider to the right.

Also, it likely needs testing against different sound systems and users as maybe it's just me having a certain volume preference or my audio card that produced a dynamic range too big to become too sensitive to slider position changes.

netnazgul commented:
Ok, now the new widget should be applied everywhere it is required. 

@abcdefg30 safe to re-test it

pchote commented on OpenRA.Mods.Common/Widgets/ExponentialSliderWidget.cs:
```diff
@@ -0,0 +1,82 @@
+			return (int)(0.5f * RenderBounds.Height + (RenderBounds.Width - RenderBounds.Height) * LinearFromExp((x - MinimumValue) / (MaximumValue - MinimumValue)));
+		}
+
+		public override void Draw()
```
This method looks to be identical to `SliderWidget.Draw` except for the tick positions.
Can you pull out a `protected IEnumerable<float> GetTickPositions()` in the base class then override it here, so that we don't need to override the drawing at all?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -123,6 +123,17 @@ static void BindSliderPref(Widget parent, string id, object group, string pref)
 			ss.OnChange += x => field.SetValue(group, x);
 		}
 
+		static void BindExpSliderPref(Widget parent, string id, object group, string pref)
```
This looks identical to `BindSliderPref` which I expect should already do what you need because `ExponentialSliderWidget` is a subclass.  This method should be able to go away.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -327,9 +338,9 @@ Action InitAudioPanel(Widget panel)
-				panel.Get<SliderWidget>("SOUND_VOLUME").OnChange += x => Game.Sound.SoundVolume = x;
-				panel.Get<SliderWidget>("MUSIC_VOLUME").OnChange += x => Game.Sound.MusicVolume = x;
-				panel.Get<SliderWidget>("VIDEO_VOLUME").OnChange += x => Game.Sound.VideoVolume = x;
+				panel.Get<ExponentialSliderWidget>("SOUND_VOLUME").OnChange += x => Game.Sound.SoundVolume = x;
```
Likewise with all these: you're not actually using the subclass properties, so it is cleaner (not to mention a much smaller diff) to access these via the parent slider class.

pchote reviewed (changes requested):
A few code simplifications, and I think there may be something wrong with the scaling constants as I find the same control issues as you say.

netnazgul commented:
Removed the unneeded binder duplicate and recasts to ExponentialSliderWidget, turned out those were overcommitments due to crashes appearing in other code.
Reverted exponential tick scaling back to linear ticks. 7 ticks represent the 60dB dynamic range (set by ExpA and ExpB values), with 1 tick span equal to 10dB.

netnazgul commented:
Aaand fixed the travis check fail.

pchote reviewed (dismissed):
LGTM. Great job!

obrakmann commented on OpenRA.Mods.Common/Widgets/ExponentialSliderWidget.cs:
```diff
@@ -0,0 +1,52 @@
+	public class ExponentialSliderWidget : SliderWidget
+	{
+		public double ExpA = 1.0e-3;
+		public double ExpB = 6.908;
```
It would be nice to have a comment here with a link to https://www.dr-lex.be/info-stuff/volumecontrols.html that explains where those numbers come from.

obrakmann reviewed (dismissed):
:+1:, just one minor thing

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/cfa49b10b78d81e64de7a4fda7de7976550e7439)
